shared-libs/lineage/test/integration.js

```diff
@@ -0,0 +1,535 @@
+        chai.expect(actual).excludingEvery('_rev').to.deep.equal({
+          _id: stub_parents._id,
+          type: stub_parents.type,
+          contact: { _id: 'something' },
+          parent: {
+            _id: 'something_else',
+            parent: dummyDoc
+          }
+        });
+      });
+    });
+
+    it('handles doc with unknown contact by leaving just the stub', function() {
+      return lineage.fetchHydratedDoc(stub_contacts._id).then(actual => {
+        chai.expect(actual).excludingEvery('_rev').to.deep.equal({
+          _id: stub_contacts._id,
+          type: stub_contacts.type,
+          contact: {
+            _id: 'something',
+            parent: dummyDoc
+          },
+          form: {}
+        });
+      });
+    });
+
+    it('handles missing contacts', function() {
+      return lineage.fetchHydratedDoc(no_contact._id).then(actual => {
+        chai.expect(actual).excluding('_rev').to.deep.equal(no_contact);
+      });
+    });
+
+    it('attaches the full lineage for reports', function() {
+      return lineage.fetchHydratedDoc(report._id).then(actual => {
+        chai.expect(actual).to.shallowDeepEqual({
+          form: 'A',
+          patient: {
+            name: report_patient.name,
+            parent: { name: report_contact.name }
+          },
+          contact: {
+            name: report_contact.name,
+            parent: {
+              name: report_parent.name,
+              contact: { phone: '+123' },
+              parent: {
+                name: report_grandparent.name,
+                contact: { phone: '+456' }
+              }
+            }
+          },
+          parent: undefined
+        });
+      });
+    });
+
+    it('attaches the contacts', function() {
+      return lineage.fetchHydratedDoc(place._id).then(actual => {
+        chai.expect(actual).to.shallowDeepEqual({
+          name: place.name,
+          contact: { phone: '+789' },
+          parent: {
+            name: place_parent.name,
+            contact: { phone: '+123' },
+            parent: {
+              name: place_grandparent.name,
+              contact: { phone: '+456' }
+            }
+          }
+        });
+      });
+    });
+
+    // This is a classic use-case: report from CHW who is the contact for their own area
+    it('attaches re-used contacts, minify handles the circular references', function() {
+      return lineage.fetchHydratedDoc(circular_report._id).then(actual => {
+        // The contact and the contact's parent's contact are the hydrated CHW
+        chai.expect(actual.contact.parent.contact.hydrated).to.equal(true);
+        chai.expect(actual.contact.hydrated).to.equal(true);
+
+        // And we can minify back to the original without error
+        lineage.minify(actual);
+        chai.expect(actual).excludingEvery('_rev').to.deep.equal(circular_report);
+      });
+    });
+
+    it('fetches contacts that it has not got via lineage', function() {
+      sinon.spy(db, 'allDocs');
```

Comment:
Why do we need `sinon` here?

Reply:
It's the only way I can see to test this functionality, as it's testing the internals of the function. The input/output would be the same for `fetch_contact` and `fetch_noContact`

Reply:
:sadface: do we benefit from testing the internals?

Reply:
Good question, it's a piece of code which is basically an optimisation so to be honest it doesn't matter too much if the optimisation does or doesn't run. I'm happy to convert it to a unit test though